linked-list.js

```diff
@@ -0,0 +1,399 @@
+      while (lastNode.next !== null) {
+        lastNode = lastNode.next;
+      }
+      lastNode.next = newNode;
+      this.tail = newNode;
+    }
+
+    return newNode;
+  }
+  arrayFromList() {
+    const array = [];
+    let lastNode = this.head;
+    while (lastNode.next !== null) {
+      array.push(lastNode.data);
+      lastNode = lastNode.next;
+    }
+    array.push(lastNode.data);
+    return array;
+  }
+  insertAfter(listNode, data) {
+    if (this.head === null || listNode === null) {
+      return null;
+    }
+
+    const newNode = new Node(data);
+
+    let lastNode = this.head;
+    while (lastNode !== null) {
+      if (lastNode === listNode) {
+        newNode.next = lastNode.next;
+        lastNode.next = newNode;
+        return newNode;
+      }
+      lastNode = lastNode.next;
+    }
+
+    return null;
+  }
+  getNode(data) {
+    if (this.head === null) {
+      return null;
+    }
+    let lastNode = this.head;
+    while (lastNode !== null) {
+      if (lastNode.data === data) {
+        return lastNode;
+      }
+      lastNode = lastNode.next;
+    }
+    return null;
+  }
+  insertBefore(listNode, data) {
+    if (this.head === null || listNode === null) {
+      return null;
+    }
+    const newNode = new Node(data);
+    let lastNode = this.head;
+    while (lastNode !== null) {
+      if (lastNode.next === listNode) {
+        newNode.next = lastNode.next;
+        lastNode.next = newNode;
+        return newNode;
+      }
+      lastNode = lastNode.next;
+    }
+    return null;
+  }
+  removeFromEnd() {
+    if (this.head === null) {
+      return;
+    }
+    if (this.head.next === null) {
+      this.head = null;
+      this.tail = null;
+      return;
+    }
+    let prevNode = null;
+    let lastNode = this.head;
+    while (lastNode.next !== null) {
+      prevNode = lastNode;
+      lastNode = lastNode.next;
+    }
+    const removedNode = prevNode.next;
+    prevNode.next = null;
+    this.tail = prevNode;
+    // eslint-disable-next-line consistent-return
+    return removedNode;
+  }
+  removeItem(data) {
+    if (this.head === null || data === null) {
+      return null;
+    }
+    if (this.head.next === null) {
+      this.head = null;
+      // eslint-disable-next-line no-param-reassign
+      this.tail = null;
+      // eslint-disable-next-line consistent-return
+      return;
+    }
+    let prevNode = null;
+    let lastNode = this.head;
+    while (lastNode !== null) {
+      if (lastNode.data === data) {
+        prevNode.next = lastNode.next;
+        // eslint-disable-next-line consistent-return
+        return lastNode;
+      }
+      prevNode = lastNode;
+      lastNode = lastNode.next;
+    }
+  }
+  filterArray(predicate) {
+    const array = [];
+    let lastNode = this.head;
+    while (lastNode !== null) {
+      if (predicate(lastNode.data)) {
+        array.push(lastNode.data);
+      }
+      lastNode = lastNode.next;
+    }
+    return array;
+  }
+  traverse(visit) {

```

Comment:
need to leave blank space between implementation of two functions

Reply:
Implemented the changes.